Fix crash in lbmap due to backendsByMapIndex containing holes

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6558)
<!-- Reviewable:end -->

tgraf committed:
lbmap: Add locking to bpfService and lbmapCache

This code has been assuming for the caller to provide appropriate protection
but it's unclear whether that is always guaranteed. Add a mutex to bpfService
and lbmapCache to protect data access.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lbmap: Avoid panic in case backendsByMapIndex contains holes

This should never happen but it may have happened. It's still unclear whether
this has happened due to a locking bug or another corruption. As the previous
commit has fixed locking, print a fat error when a hole is encountered but
avoid the panic.

Fixes: #6537

Signed-off-by: Thomas Graf <thomas@cilium.io>

vadorovsky commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -124,14 +134,24 @@ func (b *bpfService) deleteBackend(backend ServiceValue) {
+		backends[dstIndex] = b.backendsByMapIndex[i].bpfValue
+		dstIndex++
 	}
+	b.mutex.RUnlock()
```
`defer b.mutex.RUnlock()`, just for consistency?

tgraf replied:
It is faster without the defer and this is the most frequently called function

vadorovsky commented on pkg/maps/lbmap/bpfservice.go:
```diff
@@ -215,5 +241,7 @@ func (l *lbmapCache) prepareUpdate(fe ServiceKey, backends []ServiceValue) *bpfS
 func (l *lbmapCache) delete(fe ServiceKey) {
+	l.mutex.Lock()
 	delete(l.entries, fe.String())
+	l.mutex.Unlock()
```
same here

tgraf replied:
For single exit functions, there is not much of a point to pay the performance penalty of the defer

tgraf committed:
lbmap: Use length of backend map index instead of uniqueBackends map when growing

The existing code likely worked correctly but it's not logical to use the
unique backends. Fortunately, when growing beyond holes, the number of unique
backends should always be identical to the number of backend indices.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lbmap: Fix consistent load balancing when reusing backend holes

Re-use of holes in the lsit of backends can cause existing load-balancing
decisions to be changed and thus cause connection disruption. Only re-use holes
when scaling down to 0 and back up. Otherwise, the list of backends has to grow
continuously.

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm reviewed (approved):
I've looked through all code, the mutex is not needed as both structures are indeed being protected by the caller.
